### ISSUE TITLE
Bulk insert with array

### ISSUE THREAD

**Issue by obiwann** [ibmdb/node-ibm_db#489]
Hi, 

I'm trying to bulk insert multiple rows.

Example query: 

```
insert into table_name (id, name)
values (1, 'first_name'), (2, 'second_name'), (3, 'third_name')
```

I tried to find how to do this, but [this](https://www.ibm.com/developerworks/community/blogs/318b9632-5582-4635-b5fc-38a1dec81a40/entry/FAQ_Node_js_How_to_perform_bulk_insert_on_DB2_using_ibm_db_module?lang=en_us) is the only information I found so far.

I would like to bulk insert rows like it seems to be possible with mysql: ([link](https://github.com/mysqljs/mysql/pull/230)):

Example code:

```
var query = "insert into table_name (id, name) values ?";

var values = [
[1, 'first_name'],
[2, 'second_name'],
[3, 'third_name']
];

conn.query(query, [values], function (err, result) {
...
}
```

Is there a way to do this with ibm_db? When I tried, I got the "Wrong param format!" error. But maybe I'm doing something wrong?

Thanks :)

**Comment by bimalkjha** [comment]
If your server supports bulk insert, following code should work for you:
```
var query = "insert into table_name (id, name) values (1, 'first_name'), (2, 'second_name'), (3, 'third_name')";
conn.query(query, [values], function (err, result) {
...
}
```
Thanks.

**Comment by obiwann** [comment]
Hi @bimalkjha,

My server supports bulk insert and the above query works. My problem is, that I have to build the values for insertion during the app execution. 

I could do this with building the appropriate string, but this could expose the app to the sql injection. So the proper way would be, to build a query string with question marks-? (see the example code) and then insert the values in conn.query(query, values)... function. 

I'have implemented this [code](https://www.ibm.com/developerworks/community/blogs/318b9632-5582-4635-b5fc-38a1dec81a40/entry/FAQ_Node_js_How_to_perform_bulk_insert_on_DB2_using_ibm_db_module?lang=en_us) and it works (although it is not very elegant).

**Comment by utkarsh-globant** [comment]
What is the maximum limit of rows we can push in bulk insert ?
When i am trying to insert 3500 items in a go, it is giving me error :
"{ [Error: [IBM][CLI Driver] CLI0134E  Invalid parameter number. SQLSTATE=HY093]"

Even though the query is fine. 
But, if i try with around 3000 items, it works.

**Comment by geofreyflores** [comment]
Any updates on this? Looking for actual batch API methods instead of concatenating input strings manually.

**Comment by utkarsh-globant** [comment]
@geofreyflores : I guess, it directly don't support batch imports. You need to concatenate strings using looping and then execute the query. 

Only thing keep in mind is, sometimes it throws a generic error something like (invalid parameter number) even though the statement build-ed successfully, If you ever face that error, try reducing the items in bulk insert..

**Comment by davidbowenukibmcom** [comment]
We are sending batches of 600 or so using a large array.

We are concatenating many "INSERT" statements, but they all use `?` rather than risking putting the literals into the query (with all the worrying consequences that implies). I suspect it's what the example code @obiwann shared in https://github.com/ibmdb/node-ibm_db/issues/489#issuecomment-444118948 looks like, but I can't see that snippet.

It's working pretty well on the 2.5.3 version of the driver (although we have horrendous memory leaks from something when we use it). This is on node:10.

I tried 2.6.3 and the batch inserts worked for about an hour and then started failing. This even happened when we restarted the client so it seems it might be related to server state rather than client. This is on node:12.

Notably our server had less than 1,000 connections open when it failed. Contrast that to more than 1,000 when it worked fine. As a result I don't think it's connection count that is the problem.

When we have the batch insert problem we do find that sending each row individually works, but clearly the performance is appalling.

We need this to work so we can rapidly ingest data sent via an API.

Is there a plan to resolve this issue?

**Comment by bimalkjha** [comment]
@davidbowenukibmcom @shilumathai We have plan to resolve it in first half of this year. We'll start working on it soon. Thanks.

**Comment by siva-srini** [comment]
@bimalkjha Can we expect this in 2.6.5 and any approximate timeline, when we can expect.  Please provide some documentation on how to perform bulk (Array) inserts as well.  Thanks much for all your help.

**Comment by bimalkjha** [comment]
@siva-srini Yes, it will be in 2.6.5 release, working on doc update and trying to release new version before 10th June. Some other critical issues are causing delay. Thanks.

**Comment by bimalkjha** [comment]
@obiwann @p2farmer1 @utkarsh-globant @davidbowenukibmcom @siva-srini We have released ibm_db@2.7.0 with support of array insert. Please go through the documentation and test file [test-array-insert.js](https://github.com/ibmdb/node-ibm_db/blob/master/test/test-array-insert.js) to see its usage and example. Let me know if you face any issue. I am closing this issue now. Thanks.

**Comment by siva-srini** [comment]
@bimalkjha what is the limit to insert with array?  I tried with 93240 records and i am getting Segmentation fault in windows.

```
dbService.insertArray = (queryOptions) => {
  return new Promise((resolve, reject) => {
    ibmdb.open(connStr, (err, conn) => {
      if (err) reject("error connecting to database");
      conn.query(queryOptions, (err, result) => {
        if (err) conn.close(() => reject("error executing the statement. details : " + err))
        else conn.close(() => { console.log(result); resolve(result) })
      })
    });
  });
}
```
and the queryString is below
```
{
  sql: "INSERT INTO QUAL.TABLE_NAME (C1, C2, C3, C4, C5) VALUES (?, ?, ?, 'N', '2020-06-10-03.00.00.000000')",
  params: [
    { ParamType: 'ARRAY', DataType: 1, Data: [Array] },
    { ParamType: 'ARRAY', DataType: 1, Data: [Array] },
    { ParamType: 'ARRAY', DataType: 1, Data: [Array] }
  ],
  ArraySize: 93240
}
```
